### ISSUE TITLE
Instructions to migrate snikket server to a new VPS?

### ISSUE THREAD

**Issue by lispstudent** [snikket-im/snikket-server#218]
I need to help moving a small snikket server to a new VPS with a different provider.

I have never moved a docker install, and I am not sure how to do it in snikket case, where there are 4 containers:

```
# docker ps

CONTAINER ID   IMAGE                                 COMMAND                  CREATED          STATUS                    PORTS     NAMES
b1672688cb2e   snikket/snikket-server:stable         "/bin/entrypoint.sh"     59 minutes ago   Up 17 minutes (healthy)             snikket
c2f96f06a112   snikket/snikket-web-proxy:stable      "/usr/bin/tini /bin/…"   59 minutes ago   Up 17 minutes                       snikket-proxy
736db306c31f   snikket/snikket-web-portal:stable     "/bin/sh /entrypoint…"   59 minutes ago   Up 17 minutes (healthy)             snikket-portal
5699ba0c3979   snikket/snikket-cert-manager:stable   "/usr/bin/tini /bin/…"   59 minutes ago   Up 17 minutes                       snikket-certs
```

I read about `docker save` and `docker export`. What is the safest way to do this? Shall I do that for all 4 containers?

Could some instructions be added to the admin documentation, please?

**Comment by lispstudent** [comment]
Another possibiulity that came to mind, just rsync the whole docker directory?

```
docker stop $(docker ps -a -q)
systemctl stop docker.socket && systemctl stop docker.service
systemctl disable docker.socket && systemctl disable docker.service

rsync -avp --progress root@old_server_ip:/var/lib/docker/ root@new_server_ip:/var/lib/docker/

```

**Comment by mwild1** [comment]
You're right, we need some docs on this! Over in the [snikket-selfhosted](https://github.com/snikket-im/snikket-selfhosted) repo you'll find some scripts, including a backup.sh and restore.sh.

1. Before the migration, ensure your DNS record TTLs are set to something low (e.g. 60s). If they are currently set to something high, you should lower them to 60 and wait until the old records have expired.
2. Run the backup.sh on your old server and then stop it with 'docker-compose down' to make sure no further changes are made to the data.
3. Copy the backup file to your new server.
4. On the new server, follow the normal [Snikket quick-start guide](https://snikket.org/service/quickstart/) until the last step where you create the admin invite.
5. Run restore.sh on the new server, passing it the backup file you previously created. Then run 'docker compose restart' to make sure the restored data is loaded.
6. Update the IP address in your DNS records to point to the new server. Your Snikket apps should connect as if nothing happened!

After everything is confirmed working, [raise the TTLs](https://www.nslookup.io/learning/what-is-a-good-ttl-for-dns/) on your DNS records back up to something higher. This will allow the apps to (re)connect to your server more quickly during daily usage.

Give me feedback on your success following the above steps. If they seem good, I'll copy them into a documentation page. Good luck! :)

**Comment by lispstudent** [comment]
Thank you so much for the detailed instructions.

It is a bit scary for me to test, as the current Snikket VPS has about 100 online friends, and I would not want to have a long downtime if anything would go wrong.

I "dry" tested the instructions, restoring to a new VPS, but using a different domain. It all works, except of course I can verify the last "leg", .i.e. that my test account is also accessible.

I will try at around 3am or 4am tomorrow on the live VPS, and switch the DNS too, and report back.

Beside that, how could I test before changing DNS?:

Some users are across time zones and even if I revert the DNS, there could be some downtime if something goes wrong.

Is there a way, once I run `restore.sh` on the new VPS, to modify the Snikket server and proxy settings to point a temporary test domain, refresh Let's Encrypt certificates for such test domain so I could at least verify a test account is reachable?

**Comment by lispstudent** [comment]
I can confirm users had no problems. Everything worked smoothly.

Many thanks, @mwild1 !

On my side,  this ticket could be closed any time it is convenient to the maintainers.

**Comment by ip-rw** [comment]
I never really understood what problem docker volumes were trying to solve.

docker(compose)'s real charm is you can take software with the outdated habit of scattering itself all over your computer and reduce it to a single, portable directory. Bonus is you you can see what the application does at a glance, 0 noise.

Just about to setup a new Snikket install and it doesn't feel like a chore at all. Prosody is awesome, and would run on one those solar calculators or a reasonably fresh potato... 

I miss the admin commands module from prosody, I forget the name but I found myself wanting it back! Adding this in by default definitely have my vote. I'll send a pull req...

**Comment by lispstudent** [comment]
> admin commands module

I also would love to have more modules, including the admin commands.

**Comment by KageDev** [comment]
I have similar but different question:  how do I change the server domain name and keep all users data?

**Comment by mwild1** [comment]
Changing the domain name is not currently supported. The domain name is part of each user's identity.

If Alice knows someone as 'bob@hello-world.example', and then Bob changes their address to 'bob@somewhere-else.example', Alice does not know that Bob has moved, and will expect that their new address is a stranger.

Similar applies for private groups. If Bob is added to a group, when they change their address, they will no longer be able to join the group because the group does not permit their new address.

In many cases it is easiest to just start on a new domain with fresh user accounts. Some user data can be carried over using the data export in the web portal ("Manage data" section). This can be imported on the new server if you use the web registration (at the bottom of the invitation page).

We haven't ruled out supporting some kind of domain name migration support in the future, but it's not super easy.
